File: src/main/java/io/beaniejoy/dongnecafe/cafe/controller/CafeController.java
package io.beaniejoy.dongnecafe.cafe.controller;

import io.beaniejoy.dongnecafe.cafe.dto.cafe.CafeInfoResponseDto;
import io.beaniejoy.dongnecafe.cafe.dto.cafe.CafeSearchResponseDto;
import io.beaniejoy.dongnecafe.cafe.dto.cafe.CafeUpdateRequestDto;
import io.beaniejoy.dongnecafe.cafe.service.CafeService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequiredArgsConstructor
@RequestMapping("/cafes")
public class CafeController {

    private final CafeService cafeService;

    @GetMapping(value = "")
    public ResponseEntity<List<CafeSearchResponseDto>> searchCafeList(
            @PageableDefault(sort = "name", direction = Sort.Direction.ASC, size = 10) Pageable pageable) {
        List<CafeSearchResponseDto> cafeResponseList = cafeService.getCafeList(pageable);

        return ResponseEntity.ok(cafeResponseList);
    }

    @GetMapping("/{cafeId}")
    public ResponseEntity<CafeInfoResponseDto> getCafeDetailedInfo(@PathVariable("cafeId") UUID cafeId) {
        CafeInfoResponseDto cafeResponse = cafeService.getCafeInfoByCafeId(cafeId);

        return ResponseEntity.ok(cafeResponse);
    }

    @PutMapping("/{cafeId}")
    public ResponseEntity<String> updateCafeInfo(
            @PathVariable("cafeId") UUID cafeId,
            @RequestBody CafeUpdateRequestDto resource) {

        cafeService.updateCafe(cafeId, resource);

        return ResponseEntity.ok("Successfully Cafe[" + cafeId + "] Info Updated");
    }
}

File: app/service-api/src/main/kotlin/io/beaniejoy/dongnecafe/cafe/controller/CafeController.kt
package io.beaniejoy.dongnecafe.cafe.controller

import io.beaniejoy.dongnecafe.cafe.model.CafeInputDto
import io.beaniejoy.dongnecafe.cafe.model.CafeInputDtoMapper
import io.beaniejoy.dongnecafe.cafe.model.CafeOutputDto
import io.beaniejoy.dongnecafe.cafe.model.CafeOutputDtoMapper
import io.beaniejoy.dongnecafe.cafe.service.CafeService
import io.beaniejoy.dongnecafe.common.response.ApplicationResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/cafes")
class CafeController(
    private val cafeService: CafeService,
    private val cafeInputDtoMapper: CafeInputDtoMapper,
    private val cafeOutputDtoMapper: CafeOutputDtoMapper
) {
    /**
     * 신규 카페 생성
     */
    @PostMapping
    fun registerNewCafe(@RequestBody resource: CafeInputDto.RegisterCafeRequest): ApplicationResponse<CafeOutputDto.RegisteredCafe> {
        val registerCafeCommand = cafeInputDtoMapper.of(resource)

        val newCafe = cafeService.registerCafe(registerCafeCommand)

        val response = cafeOutputDtoMapper.of(newCafe)

        return ApplicationResponse
            .success("OK")
            .data(response)
    }

    /**
     * 카페 리스트 조회(검색 기능)
     */
    @GetMapping
    fun searchCafes(
        @ModelAttribute param: CafeInputDto.SearchCafesParam,
        @PageableDefault(sort = ["name"], direction = Sort.Direction.ASC, page = 0, size = 10) pageable: Pageable
    ): ApplicationResponse<Page<CafeOutputDto.CafeSearchResponse>> {
        val searchCafesQuery = cafeInputDtoMapper.of(param)

        val searchCafes = cafeService.searchCafes(
            param = searchCafesQuery,
            pageable = pageable
        )

        val responses = searchCafes.map { cafeOutputDtoMapper.of(it) }

        return ApplicationResponse
            .success()
            .data(responses)
    }

    /**
     * 단일 카페 상세 조회
     */
    @GetMapping("/{id}")
    fun getDetailedInfo(@PathVariable("id") id: Long): ApplicationResponse<CafeOutputDto.CafeDetailedResponse> {
        val cafeDetailedInfo = cafeService.getDetailedCafe(id)

        val response = cafeOutputDtoMapper.of(cafeDetailedInfo)

        return ApplicationResponse
            .success()
            .data(response)
    }

    /**
     * 단일 카페에 대한 정보 변경
     * - Cafe 기본 정보 데이터 변경
     */
    @PutMapping("/{id}")
    fun updateInfo(
        @PathVariable("id") id: Long,
        @RequestBody resource: CafeInputDto.UpdateCafeRequest
    ): ApplicationResponse<Nothing> {
        val updateCafeCommand = cafeInputDtoMapper.of(resource)

        cafeService.updateCafe(
            id = id,
            command = updateCafeCommand
        )

        return ApplicationResponse
            .success("Successfully Cafe[$id] Info Updated")
            .build()
    }
}